File: src/Compilers/CSharp/Portable/Syntax/SyntaxTriviaListBuilder.cs
// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.

using System;
using Microsoft.CodeAnalysis.CSharp.Symbols;
using Microsoft.CodeAnalysis.CSharp.Syntax;
using Microsoft.CodeAnalysis.Text;

namespace Microsoft.CodeAnalysis.CSharp.Syntax
{
    internal class SyntaxTriviaListBuilder
    {
        private SyntaxTrivia[] _nodes;
        private int _count;
        private SyntaxTrivia[] _previous;

        public SyntaxTriviaListBuilder(int size)
        {
            _nodes = new SyntaxTrivia[size];
        }

        public static SyntaxTriviaListBuilder Create()
        {
            return new SyntaxTriviaListBuilder(4);
        }

        public int Count
        {
            get { return _count; }
        }

        public void Clear()
        {
            _count = 0;
        }

        public SyntaxTrivia this[int index]
        {
            get
            {
                if (index < 0 || index > _count)
                {
                    throw new IndexOutOfRangeException();
                }

                return _nodes[index];
            }
        }

        public SyntaxTriviaListBuilder Add(SyntaxTrivia item)
        {
            if (_nodes == null || _count >= _nodes.Length)
            {
                this.Grow(_count == 0 ? 8 : _nodes.Length * 2);
            }

            _nodes[_count++] = item;
            return this;
        }

        public void Add(SyntaxTrivia[] items)
        {
            this.Add(items, 0, items.Length);
        }

        public void Add(SyntaxTrivia[] items, int offset, int length)
        {
            if (_nodes == null || _count + length > _nodes.Length)
            {
                this.Grow(_count + length);
            }

            Array.Copy(items, offset, _nodes, _count, length);
            _count += length;
        }

        public void Add(SyntaxTriviaList list)
        {
            this.Add(list, 0, list.Count);
        }

        public void Add(SyntaxTriviaList list, int offset, int length)
        {
            if (_nodes == null || _count + length > _nodes.Length)
            {
                this.Grow(_count + length);
            }

            list.CopyTo(offset, _nodes, _count, length);
            _count += length;
        }

        private void Grow(int size)
        {
            var tmp = new SyntaxTrivia[size];
            if (_previous != null)
            {
                Array.Copy(_previous, tmp, _count);
                _previous = null;
            }
            else
            {
                Array.Copy(_nodes, tmp, _nodes.Length);
            }

            _nodes = tmp;
        }

        public static implicit operator SyntaxTriviaList(SyntaxTriviaListBuilder builder)
        {
            return builder.ToList();
        }

        public SyntaxTriviaList ToList()
        {
            if (_count > 0)
            {
                if (_previous != null)
                {
                    this.Grow(_count);
                }

                switch (_count)
                {
                    case 1:
                        return new SyntaxTriviaList(default(SyntaxToken), _nodes[0].UnderlyingNode, position: 0, index: 0);
                    case 2:
                        return new SyntaxTriviaList(default(SyntaxToken), Syntax.InternalSyntax.SyntaxList.List(
                            (Syntax.InternalSyntax.CSharpSyntaxNode)_nodes[0].UnderlyingNode,
                            (Syntax.InternalSyntax.CSharpSyntaxNode)_nodes[1].UnderlyingNode), position: 0, index: 0);
                    case 3:
                        return new SyntaxTriviaList(default(SyntaxToken),
                            Syntax.InternalSyntax.SyntaxList.List(
                                (Syntax.InternalSyntax.CSharpSyntaxNode)_nodes[0].UnderlyingNode,
                                (Syntax.InternalSyntax.CSharpSyntaxNode)_nodes[1].UnderlyingNode,
                                (Syntax.InternalSyntax.CSharpSyntaxNode)_nodes[2].UnderlyingNode),
                            position: 0, index: 0);
                    default:
                        {
                            var tmp = new ArrayElement<Syntax.InternalSyntax.CSharpSyntaxNode>[_count];
                            for (int i = 0; i < _count; i++)
                            {
                                tmp[i].Value = (Syntax.InternalSyntax.CSharpSyntaxNode)_nodes[i].UnderlyingNode;
                            }

                            return new SyntaxTriviaList(default(SyntaxToken), Syntax.InternalSyntax.SyntaxList.List(tmp), position: 0, index: 0);
                        }
                }
            }
            else
            {
                return default(SyntaxTriviaList);
            }
        }
    }
}

File: src/Compilers/VisualBasic/Portable/Syntax/SyntaxTriviaListBuilder.vb
' Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.

Imports System.Collections
Imports System.Collections.Generic
Imports Microsoft.CodeAnalysis.Text
Imports Microsoft.CodeAnalysis.VisualBasic.Symbols
Imports Microsoft.CodeAnalysis.VisualBasic.Syntax

Namespace Microsoft.CodeAnalysis.VisualBasic.Syntax

    Friend Class SyntaxTriviaListBuilder
        Private _count As Integer
        Private _nodes As SyntaxTrivia()

        Public ReadOnly Property Count As Integer
            Get
                Return _count
            End Get
        End Property

        Public Sub New(size As Integer)
            _nodes = New SyntaxTrivia(size - 1) {}
        End Sub

        Default Public ReadOnly Property Item(index As Integer) As SyntaxTrivia
            Get
                Debug.Assert(index >= 0)
                Debug.Assert(index < _count)

                Return _nodes(index)
            End Get
        End Property

        Public Sub Add(list As SyntaxTriviaList)
            Add(list, 0, list.Count)
        End Sub

        Public Sub Add(items As SyntaxTrivia())
            Add(items, 0, items.Length)
        End Sub

        Public Function Add(item As SyntaxTrivia) As SyntaxTriviaListBuilder
            If ((_nodes Is Nothing) OrElse (_count >= _nodes.Length)) Then
                Grow(If((_count = 0), 8, (_nodes.Length * 2)))
            End If
            _nodes(_count) = item
            _count += 1
            Return Me
        End Function

        Public Sub Add(items As SyntaxTrivia(), sourceOffset As Integer, length As Integer)
            If ((_nodes Is Nothing) OrElse ((_count + length) > _nodes.Length)) Then
                Grow((_count + length))
            End If
            Array.Copy(items, sourceOffset, _nodes, _count, length)
            _count = (_count + length)
        End Sub

        Public Sub Add(list As SyntaxTriviaList, sourceOffset As Integer, length As Integer)
            If ((_nodes Is Nothing) OrElse ((_count + length) > _nodes.Length)) Then
                Grow((_count + length))
            End If
            list.CopyTo(sourceOffset, _nodes, _count, length)
            _count = (_count + length)
        End Sub

        Public Sub Clear()
            _count = 0
        End Sub

        Public Shared Function Create() As SyntaxTriviaListBuilder
            Return New SyntaxTriviaListBuilder(4)
        End Function

        Private Sub Grow(size As Integer)
            Dim tmp As SyntaxTrivia() = New SyntaxTrivia(size - 1) {}
            Array.Copy(_nodes, tmp, _nodes.Length)
            _nodes = tmp
        End Sub

        Public Shared Widening Operator CType(builder As SyntaxTriviaListBuilder) As SyntaxTriviaList
            Return builder.ToList
        End Operator

        Public Function ToList() As SyntaxTriviaList
            If (_count <= 0) Then
                Return New SyntaxTriviaList
            End If
            Select Case _count
                Case 1
                    Return New SyntaxTriviaList(Nothing, _nodes(0).UnderlyingNode, 0, 0)
                Case 2
                    Return New SyntaxTriviaList(Nothing, InternalSyntax.SyntaxList.List(DirectCast(_nodes(0).UnderlyingNode, InternalSyntax.VisualBasicSyntaxNode), DirectCast(_nodes(1).UnderlyingNode, InternalSyntax.VisualBasicSyntaxNode)), 0, 0)
                Case 3
                    Return New SyntaxTriviaList(Nothing, InternalSyntax.SyntaxList.List(DirectCast(_nodes(0).UnderlyingNode, InternalSyntax.VisualBasicSyntaxNode), DirectCast(_nodes(1).UnderlyingNode, InternalSyntax.VisualBasicSyntaxNode), DirectCast(_nodes(2).UnderlyingNode, InternalSyntax.VisualBasicSyntaxNode)), 0, 0)
            End Select
            Dim tmp = New ArrayElement(Of InternalSyntax.VisualBasicSyntaxNode)(_count - 1) {}
            Dim i As Integer
            For i = 0 To _count - 1
                tmp(i).Value = DirectCast(_nodes(i).UnderlyingNode, InternalSyntax.VisualBasicSyntaxNode)
            Next i
            Return New SyntaxTriviaList(Nothing, InternalSyntax.SyntaxList.List(tmp), 0, 0)
        End Function

    End Class

End Namespace
